src/emulator/storage/files.ts

```diff
@@ -291,11 +295,28 @@ export class StorageLayer {
     const file = new StoredFile(finalMetadata, filePath);
     this._files.set(filePath, file);
 
+    return { upload, file };
+  }
+
+  public persistUpload(finalizedUpload: FinalizedUpload): void {
+    const { upload, file } = finalizedUpload;
+    const filePath = this.path(upload.bucketId, upload.objectId);
+
     this._persistence.deleteFile(filePath, true);
     this._persistence.renameFile(upload.fileLocation, filePath);
 
     this._cloudFunctions.dispatch("finalize", new CloudStorageObjectMetadata(file.metadata));
-    return { upload: upload, file: file };
+  }
+
+  public finalizeUpload(uploadId: string): FinalizedUpload | undefined {
```

Comment:
nit: Method names should be descriptive of their behavior, not descriptors of their callsites. Litmus test is asking yourself if you understand what the method's purpose is from just the name. In this case, `finishUpload` and `finalizeUpload` are both ambiguous. See go/tsstyle#descriptive-names

Reply:
Ok. I'll get rid of `finalizeUpload` altogether since we're exposing the two methods it's calling and there's only one instance of its usage.